src/fp32.rs

```diff
@@ -0,0 +1,293 @@
+        let result = (self.p as u64) * (w as u64);
+        let hi = hi32(result);
+        let lo = lo32(result) as u32;
+        let (result, carry) = zz[0].overflowing_add(lo);
+        zz[0] = result;
+        let result = zz[1] as u64 + hi + carry as u64;
+        zz[1] = lo32(result) as u32;
+        let cc = hi32(result) as u32;
+
+        // z = (z1)
+        let prod = zz[1];
+
+        // Final subtraction
+        // If z >= p, then z = z - p
+
+        //    cc, z
+        // -   0, p
+        // ========
+        // b1,s1,s0
+        let (s0, b0) = prod.overflowing_sub(self.p);
+        let (_s1, b1) = cc.overflowing_sub(b0 as u32);
+        // if b1 == 1: return z
+        // else:       return s0
+        let mask = 0u32.wrapping_sub(b1 as u32);
+        (prod & mask) | (s0 & !mask)
+    }
+
+    /// Modular exponentiation, i.e., `x^exp (mod p)` where `p` is the modulus. Note that the
+    /// runtime of this algorithm is linear in the bit length of `exp`.
+    pub fn pow(&self, x: u32, exp: u32) -> u32 {
+        let mut t = self.montgomery(1);
+        for i in (0..32 - exp.leading_zeros()).rev() {
+            t = self.mul(t, t);
+            if (exp >> i) & 1 != 0 {
+                t = self.mul(t, x);
+            }
+        }
+        t
+    }
+
+    /// Modular inversion, i.e., x^-1 (mod p) where `p` is the modulus. Note that the runtime of
+    /// this algorithm is linear in the bit length of `p`.
+    #[inline(always)]
+    pub fn inv(&self, x: u32) -> u32 {
+        self.pow(x, self.p - 2)
+    }
+
+    /// Negation, i.e., `-x (mod p)` where `p` is the modulus.
+    #[inline(always)]
+    pub fn neg(&self, x: u32) -> u32 {
+        self.sub(0, x)
+    }
+
+    /// Maps an integer to its internal representation. Field elements are mapped to the Montgomery
+    /// domain in order to carry out field arithmetic. The result will be in [0, p).
+    ///
+    /// # Example usage
+    /// ```text
+    /// let integer = 1; // Standard integer representation
+    /// let elem = fp.montgomery(integer); // Internal representation in the Montgomery domain
+    /// assert_eq!(elem, 1048575);
+    /// ```
+    #[inline(always)]
+    pub fn montgomery(&self, x: u32) -> u32 {
+        modp(self.mul(x, self.r2), self.p)
+    }
+
+    /// Maps a field element to its representation as an integer. The result will be in [0, p).
+    ///
+    /// #Example usage
+    /// ```text
+    /// let elem = 1048575; // Internal representation in the Montgomery domain
+    /// let integer = fp.residue(elem); // Standard integer representation
+    /// assert_eq!(integer, 1);
+    /// ```
+    #[inline(always)]
+    pub fn residue(&self, x: u32) -> u32 {
+        modp(self.mul(x, 1), self.p)
+    }
+}
+
+#[inline(always)]
+fn modp(x: u32, p: u32) -> u32 {
+    let (z, carry) = x.overflowing_sub(p);
+    let m = 0u32.wrapping_sub(carry as u32);
+    z.wrapping_add(m & p)
+}
+
+pub(crate) const FP32: FieldParameters32 = FieldParameters32 {
```

Comment:
Could you check in a corresponding change to `documentation/field_parameters.sage` to generate these values?

Reply:
I've added a field in the parameters structure, let me know what do you think.